lib/libxdp/xsk.c

```diff
@@ -683,6 +713,60 @@ static struct xdp_program *xsk_lookup_program(int ifindex)
 	return prog;
 }
 
+static int xsk_update_prog_refcnt(int refcnt_map_fd, int delta)
+{
+	struct bpf_map_info map_info = {};
+	__u32 info_len = sizeof(map_info);
+	__u32 refcnt_offset = 0, key = 0;
+	int lock_fd, ret, *value;
+	void *value_data = NULL;
+
+	lock_fd = xdp_lock_acquire();
+	if (lock_fd < 0)
+		return lock_fd;
+
+	ret = bpf_obj_get_info_by_fd(refcnt_map_fd, &map_info, &info_len);
+	if (ret)
+		goto unlock;
+
+	value_data = malloc(map_info.value_size);
+	if (!value_data) {
+		ret = -ENOMEM;
+		goto unlock;
+	}
+
+	ret = bpf_map_lookup_elem(refcnt_map_fd, &key, value_data);
+	if (ret)
+		goto unlock;
+
+	value = value_data + refcnt_offset;
+
+	/* If refcount is 0, program is awaiting detach and can't be used */
+	if (*value) {
+		*value += delta;
+		ret = bpf_map_update_elem(refcnt_map_fd, &key, value_data, 0);
+		if (ret)
+			goto unlock;
+	}
+
+	ret = *value;
```

Comment:
Shouldn't the retval be a negative error code if the `if (*value)` check fails above?